### PR TITLE
feat: Add a backend for replicated sharing for rings, as well as code of casting between different rings and fields.

### PR DESCRIPTION
**Pull request by rw0x0**

### CODE REVIEW

**Comment by dkales** [comment]
IDK how to easiest fix this now since the first 2 commits were already in the other PR?
probably just cherry-picking the other ones onto main?

**Comment by rw0x0** [comment]
I just rebased everything, worked out nicely

File: mpc-core/src/protocols/rep3_ring.rs [hunk 1]
```diff
@@ -2,16 +2,107 @@
...
+    let a = rng.gen::<RingElement<T>>();
+    let b = rng.gen::<RingElement<T>>();
+    let c = val ^ a ^ b;
+    let share1 = Rep3RingShare::new_ring(a, c);
+    let share2 = Rep3RingShare::new_ring(b, a);
+    let share3 = Rep3RingShare::new_ring(c, b);
+    [share1, share2, share3]
+}
+
+//TODO RENAME ME TO COMBINE_ARITHMETIC_SHARE
```
Comment (dkales) [inline mpc-core/src/protocols/rep3_ring.rs:74]: do we do this rename here?

File: mpc-core/src/protocols/rep3_ring.rs [hunk 2]
```diff
@@ -2,16 +2,107 @@
...
+) -> Vec<RingElement<T>> {
+    assert_eq!(share1.len(), share2.len());
+    assert_eq!(share2.len(), share3.len());
+
+    itertools::multizip((share1.into_iter(), share2.into_iter(), share3.into_iter()))
+        .map(|(x1, x2, x3)| x1.a + x2.a + x3.a)
+        .collect::<Vec<_>>()
+}
+
+//TODO RENAME ME TO COMBINE_BINARY_SHARE
```
Comment (dkales) [inline mpc-core/src/protocols/rep3_ring.rs:100]: this is leftover?

File: mpc-core/src/protocols/rep3_ring.rs [hunk 3]
```diff
@@ -2,16 +2,107 @@
...
+    share2: Rep3RingShare<T>,
+    share3: Rep3RingShare<T>,
+) -> RingElement<T> {
+    share1.a + share2.a + share3.a
+}
+
+/// Reconstructs a vector of ring elements from its arithmetic replicated shares.
+/// # Panics
+/// Panics if the provided `Vec` sizes do not match.
+pub fn combine_ring_elements<T: IntRing2k>(
```
Comment (dkales) [inline mpc-core/src/protocols/rep3_ring.rs:87]: do we need a vec variant for binary as well?

File: mpc-core/src/protocols/rep3_ring.rs [hunk 4]
```diff
@@ -2,16 +2,107 @@
...
+    share1: Vec<Rep3RingShare<T>>,
```
Comment (dkales) [inline mpc-core/src/protocols/rep3_ring.rs:88]: probably can take slices here, no?

File: mpc-core/src/protocols/rep3_ring/casts.rs [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+where
+    T: IntRing2k + AsPrimitive<U>,
+    U: IntRing2k,
+    N: Rep3Network,
+    Standard: Distribution<T> + Distribution<U>,
+{
+    match io_context.a2b_type {
+        A2BType::Direct => cast_a2b(x, io_context),
+        A2BType::Yao => cast_gc(x, io_context),
+        A2BType::StreamingYao => {
```
Comment (dkales) [inline mpc-core/src/protocols/rep3_ring/casts.rs:40]: should we just rip out streaming yao?

File: mpc-core/src/protocols/rep3_ring/ring/int_ring.rs [hunk 6]
```diff
@@ -38,18 +43,36 @@ pub trait IntRing2k:
...
     fn write<W: std::io::Write>(&self, writer: W) -> IoResult<()>;
 
     /// Returns the effective number of bits (i.e., how many LSBs are set)
     fn bits(&self) -> usize;
 
+    /// Casts this type to a BigUint
+    fn cast_to_biguint(&self) -> BigUint;
+
+    /// Casts a BigUint to this type
+    fn cast_from_biguint(biguint: &BigUint) -> Self;
```
Comment (dkales) [inline mpc-core/src/protocols/rep3_ring/ring/int_ring.rs:74]: what happens when this is larger than the type? should this not return a result?

File: mpc-core/src/protocols/rep3_ring/yao.rs [hunk 7]
```diff
@@ -306,3 +324,367 @@ pub fn joint_input_binary_xored<T: IntRing2k, N: Rep3Network>(
...
+                ))?;
+            }
+        }
+    }
+
+    Ok(res)
+}
+
+/// A cast of a vector of Rep3PrimeFieldShare to a vector of Rep3RingShare
+pub fn field_to_ring_many<F: PrimeField, T: IntRing2k, N: Rep3Network>(
```
Comment (dkales) [inline mpc-core/src/protocols/rep3_ring/yao.rs:449]: what does this do if input > ring_size? just mod 2^K?

File: tests/tests/mpc/rep3_ring.rs [hunk 8]
```diff
@@ -0,0 +1,1627 @@
...
+            )*
+        };
+        ($expr:ident,[$($t1:ty),*],$t2:tt) => {
+            $(
+                apply_to_all2!($expr,[$t1], $t2);
+            )*
+        };
+    }
+
+    // TODO we dont need channels, we can just join
```
Comment (dkales) [inline tests/tests/mpc/rep3_ring.rs:52]: should we just do this then?

File: mpc-core/src/protocols/rep3_ring/ring/int_ring.rs (hunk 6 above)
Comment (rw0x0) [reply mpc-core/src/protocols/rep3_ring/ring/int_ring.rs:74]: is really meant as a downcast that cuts off stuff

File: mpc-core/src/protocols/rep3_ring.rs (hunk 1 above)
Comment (rw0x0) [reply mpc-core/src/protocols/rep3_ring.rs:74]: This is copied from the original rep3, don't know who added that comment...

File: mpc-core/src/protocols/rep3_ring.rs (hunk 2 above)
Comment (rw0x0) [reply mpc-core/src/protocols/rep3_ring.rs:100]: This is copied from the original rep3, don't know who added that comment...

File: mpc-core/src/protocols/rep3_ring.rs (hunk 1 above)
Comment (rw0x0) [reply mpc-core/src/protocols/rep3_ring.rs]: I removed it

File: mpc-core/src/protocols/rep3_ring.rs (hunk 4 above)
Comment (rw0x0) [reply mpc-core/src/protocols/rep3_ring.rs]: adapted (also in the old rep3)

File: mpc-core/src/protocols/rep3_ring.rs (hunk 2 above)
Comment (rw0x0) [reply mpc-core/src/protocols/rep3_ring.rs]: I removed it

File: mpc-core/src/protocols/rep3_ring/ring/int_ring.rs (hunk 6 above)
Comment (dkales) [reply mpc-core/src/protocols/rep3_ring/ring/int_ring.rs:74]: maybe add a comment then

Comment (rw0x0) [reply mpc-core/src/protocols/rep3_ring/ring/int_ring.rs:74]: done

**Review by dkales — APPROVED** [review]

File: mpc-core/src/protocols/rep3_ring/yao.rs (hunk 7 above)
Comment (rw0x0) [reply mpc-core/src/protocols/rep3_ring/yao.rs:449]: truncation, have added a comment

File: mpc-core/src/protocols/rep3_ring/casts.rs (hunk 5 above)
Comment (rw0x0) [reply mpc-core/src/protocols/rep3_ring/casts.rs]: have done so, but kept the backend implementaiton

**Pull request merged**